crates/subspace-block-relay/src/lib.rs

```diff
@@ -90,45 +92,42 @@ pub(crate) struct Resolved<ProtocolUnitId, ProtocolUnit> {
 #[async_trait]
 pub(crate) trait ProtocolClient<DownloadUnitId, ProtocolUnitId, ProtocolUnit>:
     Send + Sync
-where
-    DownloadUnitId: Encode + Decode,
-    ProtocolUnitId: Encode + Decode,
-    ProtocolUnit: Encode + Decode,
 {
+    type ProtocolReq: Send + Sync + Encode + Decode;
+    type ProtocolRsp: Send + Sync + Encode + Decode;
+
     /// Builds the protocol portion of the initial request
-    fn build_initial_request(&self) -> Option<Vec<u8>>;
+    fn build_initial_request(&self) -> Self::ProtocolReq;
 
-    /// Resolves the initial response to produce the protocol units
-    async fn resolve(
+    /// Resolves the initial response to produce the protocol units.
+    /// `encoder_fn` needs to be called to generate the request payload
+    /// as part of request/response sequences initiated by the protocol.
+    async fn resolve_initial_response(
         &self,
-        response: Vec<u8>,
-        stub: Arc<dyn NetworkStub>,
+        response: Self::ProtocolRsp,
+        encoder_fn: EncoderFn<Self::ProtocolReq>,
+        network: Arc<dyn NetworkInterface>,
     ) -> Result<(DownloadUnitId, Vec<Resolved<ProtocolUnitId, ProtocolUnit>>), RelayError>;
```

Comment:
Why `encoder_fn` and not request itself with bound `Request: Encode`?

Reply:
The encoder_fn just wraps the protocol message in `ServerMessage::ProtocolReq`, as this hits the same server on the request/response side as the initial request

Reply:
I understand, but that wrapper is the same all the time, why not take `request` as an input and calling `ServerMessage::ProtocolReq(request).encode()` internally. It doesn't look nice that you create a callback, box it, all just to call predictable `.encode()` method.

Reply:
Nope, that would be a leaky abstraction. Few things:
1. If we could run two request/response server end points, we could potentially send the client and protocol messages through completely different paths. But since we have one server, we need to mux/demux using the `ServerMessage` outer wrapper
2. The protocol should not know about the client messages (it could be consensus/execution/something else), it is embedded/managed in the client. 
3. The server message may be completely different between clients. 

So the wrapping is abstracted from the protocol and done by the client